internal/pg/postgres/password/md5.go

```diff
@@ -0,0 +1,72 @@
+package password
+
+/*
+ Copyright 2020 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"fmt"
+)
+
+// md5Prefix is used as a prefix to the hashed password, i.e. "md5[a-f0-9]{32}"
+const md5Prefix = "md5"
+
+var (
+	// ErrMD5PasswordInvalid is returned when the password attributes are invalid
+	ErrMD5PasswordInvalid = errors.New(`invalid password attributes. must provide "username" and "password"`)
+)
+
+// MD5Password implements the PostgresPassword interface for hashing passwords
+// using the PostgreSQL MD5 method
+type MD5Password struct {
+	// password is the plaintext password
+	password string
+	// username is the PostgreSQL username
+	username string
+}
+
+// Build creates the MD5 password format for PostgreSQL which resembles
+// "md5" + md5("password" + "username")
+func (m *MD5Password) Build() (string, error) {
+	// create the plaintext password/salt that PostgreSQL expects as a byte string
+	plaintext := []byte(fmt.Sprintf("%s%s", m.password, m.username))
+	// set up the password hasher
+	hasher := md5.New()
+	// add the above plaintext to the hash
+	hasher.Write(plaintext)
+	// finish the transformation by getting the string value of the MD5 hash and
+	// encoding it in hexadecimal for PostgreSQL, appending "md5" to the front
+	return fmt.Sprintf("md5%s", hex.EncodeToString(hasher.Sum(nil))), nil
+}
+
+// setAttributes accepts two values in this order in order to prep the
+// MD5Password:
+//
+// username: the PostgreSQL username
+// password: the plaintext password
+func (m *MD5Password) setAttributes(attr ...string) error {
+	// if there is anything other than two attributes, return an error
+	if len(attr) != 2 {
+		return ErrMD5PasswordInvalid
+	}
+
+	// otherwise, set the username and password
+	m.username = attr[0]
+	m.password = attr[1]
+
+	return nil
+}
```

Comment:
This seems like it wants to be a constructor but it's also unexported. The caller can instead:

```go
MD5Password{username: "x", password: "y"}
```